### PR TITLE
Add: installation, configuration, and initial tests with Cypress

### PR DESCRIPTION
**Pull request by AnthonyMadia**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- Dependencies:
    - `cypress`
    - `@testing-library/cypress`
    - `start-server-and-test`
        - allows you to run only one script to get cypress running rather than running yarn start to get localhost:3000 running and then yarn cypress open
- Add `yarn test:e2e` to CircleCI to ensure CI runs with Cypress
- Add Cypress dirs to  `.gitignore`
- Mock network requests with `cy.intercept`
- Tests for:
    - `landing.cy.test`
    - `signup.cy.test`
    - `login.cy.test`

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by AnthonyMadia** [comment]
The problem I want to fix before merging this is to mock the backend correctly. Currently, I have some example data in the `fixtures` directory which can mock the JSON reponses.

Also, the `Cypress.on('uncaught:exception'`... line makes the test still run when encountering uncaught errors; you can still see the errors on the GUI with this line. 

Aside from this, everything seems to be working well with the config of Cypress but want to understand exactly how the backend requests are being made and get them working correctly to signup and login a user.

To repllicate the errors locally, check out this PR, and run `yarn test:e2e:dev` . This will fire up [[localhost:3000](http://localhost:3000/)](http://localhost:3000) and run the Cypress GUI automatically.

**Comment by vimwitch** [comment]
I think a good approach for mocking the backend will be running a proxy between the webapp and an ethereum node and record the requests and responses as synchronization happens. Those should be serialized to json and replayed as requests are made during the tests. This should make it easy enough to simulate a unirep instance. [This](https://github.com/vimwitch/hardhelmet/blob/main/index.js) codebase should be a decent starting point for this approach.

We're about to change how the contracts work. Right now we have to pull a bunch of events and process them, but in the future we'll be able to pull the state via a single (or a few) requests. It will probably be 2-3 weeks before these changes are reflected in the frontend codebase, so it's up to you how much effort you want to put into mocking.

**Comment by AnthonyMadia** [comment]
@vimwitch For clarity, the `cy.intercept()` method [stubs](https://docs.cypress.io/api/commands/intercept#Syntax) network requests and responses so I think a proxy would be unnecessary and add complexity. The issue I was having with the backend was understanding the response from the requests so that they can be stubbed correctly, specifically around when a user signs up and logs in. 

For example, when the loading button is clicked for logging in a user, a request is sent to the `/api/genInvitationCode/...` endpoint, but because I do not have the correct JSON response in the `fixtures/genInvitationCode.json` directory, it [fails](https://github.com/Unirep/Unirep-Social/pull/55/files#:~:text=cy.get(%27*%5Bclass%5E%3D%22loading,5.6.2)%20error%20thrown%20here). So, my thinking is that if the proper JSON is put in the fixtures directory, then the login test would work.

>We're about to change how the contracts work. Right now we have to pull a bunch of events and process them, but in the future we'll be able to pull the state via a single (or a few) requests. It will probably be 2-3 weeks before these changes are reflected in the frontend codebase, so it's up to you how much effort you want to put into mocking.

With these changes being made, I do not think it would interrupt anything because I can simply stub them using the intercept method. The only thing that would probably have to change will be the response being sent back from the request on the intercept method.

**Comment by AnthonyMadia** [comment]
More context on network requests with Cypress [here](https://docs.cypress.io/guides/guides/network-requests#Use-Server-Responses).

**Comment by vimwitch** [comment]
Ah yea i didn't mean use the proxy during the test, but rather when you're using the app. So for the signup/login you could just do it in the browser and look at the json response.

**Comment by AnthonyMadia** [comment]
Ah okay I see, I will try that. Was using chrome dev tools with Cypress to look at network

**Comment by AnthonyMadia** [comment]
So, I am trying a few different things to get these tests to login/signup a user and when I changed the body to include the `unirepAddress` and `unirepSocialAddress` to match the backend [controller](https://github.com/Unirep/Unirep-Social/blob/main/packages/backend/src/routes/ConfigRouter.ts#:~:text=UNIREP_SOCIAL%20%7D%20from%20%27../constants%27-,export%20default%20(app%3A%20Express)%20%3D%3E%20%7B,%7D,-Footer) and I get a different error for the login test: `Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.6.8)`. Not sure exactly where this is stemming from though. 

Currently reviewing the genInvitation code [controllers](https://github.com/Unirep/Unirep-Social/blob/main/packages/backend/src/routes/GenInvitationCodeRouter.ts#:~:text=%7D-,async%20function%20generateCode(req%2C%20res)%20%7B,%7D,-Footer) in the backend.

I know auth may be added and these pages may change, but still very confused as to why the login/signup processes are not working.

**Comment by AnthonyMadia** [comment]
@vimwitch Do you mind taking a look at logging in a user on your machine when you have some time? Want to make sure I am not missing anything or if there's an issue with my machine. After the login finally works, I can create an abstraction for it and e2e test the rest of the app.

**Comment by vimwitch** [comment]
Yep, I'll record some stuff later today. The responses should be pretty simple.

**Comment by AnthonyMadia** [comment]
That would be helpful!

**Comment by AnthonyMadia** [comment]
I think the way the tests are now give a solid foundation for when auth is added. When auth is added the tests will most likely fail, but can be fixed by changing the parent signup command in `support/commands.ts` to something like [this](https://docs.cypress.io/guides/end-to-end-testing/auth0-authentication#:~:text=programmatically%20login%20into%20Auth0%2C) when the time comes depending on how auth is implemented. 

Also, made a few more optimizations such as moving repeated code from the deploy and signup processes into the `comands.ts` file and using a `before` instead of a `beforeEach` for the signup processes to reduce runtime of tests. Still looking for more optimizations, but might make sense to merge this. @vimwitch

**Review by vimwitch — APPROVED** [review]
Agreed, this is a great start for e2e testing

File: .circleci/config.yml [hunk 1]
```diff
@@ -102,4 +102,4 @@ jobs:
                   command: yarn build --scope @unirep-social/core --scope @unirep-social/frontend
             - run:
                   name: Test
-                  command: cd packages/frontend && yarn test
```
[the remark this answers is not in the document]
Comment (AnthonyMadia) [reply .circleci/config.yml:105]: Will keep this in mind.

[the remark this answers is not in the document]
Comment (vimwitch) [reply .circleci/config.yml:105]: I already did this in this pr :+1:

[the remark this answers is not in the document]
Comment (AnthonyMadia) [reply .circleci/config.yml:105]: Yes, was just reviewing and could have sworn the change was made haha

**Pull request merged**